Update to latest containerd, runc, docker

cbgbt opened the pull request:
**Issue number:** #1971 



**Description of changes:** This moves Bottlerocket to the latest versions of containerd, runc, and docker.


**Testing done:**
Our own changes are pretty small here, but testing for feature compatibility, performance, and regressions is critical. All testing has been done with a local build of the [latest changes to the bottlerocket-sdk](https://github.com/bottlerocket-os/bottlerocket-sdk/commit/907998cf3f8a4093868042ba15fea8cf7c0db278).

* [x] Basic testing of aws-ecs-1 variant
* [x] Basic testing of aws-k8s-1.21 variant
* [x] Sonobuoy conformance testing on several variants
* [x] Test containerd Pod pull improvements from [this containerd improvement](https://github.com/containerd/containerd/pull/6816).
* [x] Test containerd-related settings for compatbility. In particular, #1963 
* [x] Test rlimit patch rebase



**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

cbgbt committed:
containerd: update to 1.6.4

This includes a rebase of the rlimit patch.

cbgbt committed:
runc: update to 1.1.2

cbgbt committed:
docker: update to 20.10.16

Includes changes for both docker-engine and docker-cli.

bcressey reviewed (commented):
LGTM pending testing. I'll also pull these commits into my own development branch to give them more exercise.

We should confirm that `ulimit -n` in a pod returns 65k. It'd also be good to keep an eye on the [runc 1.1.3 milestone](https://github.com/opencontainers/runc/milestone/19?closed=1) to see if any regressions appear. The dbus fix there might be worth pulling in.

cbgbt commented:
> LGTM pending testing. I'll also pull these commits into my own development branch to give them more exercise.

Thanks!

> We should confirm that `ulimit -n` in a pod returns 65k.

I've confirmed that the ulimit is appropriately set inside a kubernetes Pod running in this build of containerd!

> It'd also be good to keep an eye on the [runc 1.1.3 milestone](https://github.com/opencontainers/runc/milestone/19?closed=1) to see if any regressions appear. The dbus fix there might be worth pulling in.

Good tip. A lot of the existing testing for this build is in-flight, I'll complete it and then look into re-running some subset of testing with a patch for that runc issue in place.

cbgbt commented:
I've moved testing performance impact of this change to https://github.com/bottlerocket-os/bottlerocket/issues/2171, which is also added to the 1.8.0 milestone and is meant to include changes from https://github.com/bottlerocket-os/bottlerocket/pull/2166.

I'm merging in and testing the runc/dbus fix, then this should be ready for review.

cbgbt commented:
It turns out that https://github.com/opencontainers/runc/pull/3476 doesn't totally resolve the issue with dbus.

Our Kubernetes build relies on a vendored version of runc for `libct`, so we need to also modify that packaging in order to include the fix. This issue is tracked upstream in Kubernetes in https://github.com/kubernetes/kubernetes/issues/100328

I've opened https://github.com/bottlerocket-os/bottlerocket/issues/2172 to track the issue.
